tour_model_eval/confirmed_trips_eval_bins_clusters.py

```diff
@@ -20,54 +22,213 @@
 map_pur_dict = {'course':'school','work_- lunch break':'lunch_break','on_the way home':'home',
                'insurance_payment':'insurance'}
 
-# precision_bins takes five parameters
-# - all_bins_preci: the list that collects precision of each bin, should pass in an empty list
-# - sp2en=None means no need to translate language
-#   sp2en='True' will use span_eng_dict to change Spanish to English
-#
-# - cvt_purpose=None means no need to convert purposes
-#   cvt_purpose='True' will use map_pur_dict to convert purposes
-#   using this parameter should also set sp2en='True'
-def precision_bins (all_bins_preci,bins,non_empty_trips,sp2en=None,cvt_purpose=None):
-    for bin in bins:
-        bin_user_input = (non_empty_trips[i].data["user_input"] for i in bin if
-                          non_empty_trips[i].data["user_input"] != {})
-        bin_df = pd.DataFrame(data=bin_user_input)
-        if sp2en == 'True':
-            bin_df = bin_df.replace(span_eng_dict)
-        if cvt_purpose == 'True':
-            bin_df = bin_df.replace(map_pur_dict)
-        duplic_trips = bin_df[bin_df.duplicated(keep=False)]
-
-        # for bin that doesn't have duplicate trips, assign 0 as precision
-        if duplic_trips.empty and len(bin_df) > 1:
-            all_bins_preci.append(0)
-        # for bin only has one trip, assign 1.0 as precision
-        elif len(bin_df) == 1:
-            all_bins_preci.append(1.0)
-        else:
-            duplic = duplic_trips.groupby(duplic_trips.columns.tolist()).apply(lambda x: tuple(x.index)).tolist()
-            max_duplic = max(duplic, key=lambda i: len(i))
-            precision = round(len(max_duplic) / len(bin), 2)
-            all_bins_preci.append(precision)
-    return all_bins_preci
-
-
-# precision_all_users takes four parameters
-# - all_users: pass in all participants' data
-# - sp2en: default None, no need to change language
-# - cvt_purpose: default None, no need to convert purpose
-def precision_bin_all_users(all_users,radius,sp2en=None,cvt_purpose=None):
-    all_users_preci = []
+def get_user_ls(all_users):
+    user_ls = []
+    for i in range(len(all_users)):
+        curr_user = 'user' + str(i + 1)
+        user_ls.append(curr_user)
+    return user_ls
+
+
+def filter_data(user,radius):
+    trips = pipeline.read_data(uuid=user, key=esda.CONFIRMED_TRIP_KEY)
+    non_empty_trips = [t for t in trips if t["data"]["user_input"] != {}]
+    non_empty_trips_df = pd.DataFrame(t["data"]["user_input"] for t in non_empty_trips)
+    valid_trips_df = non_empty_trips_df.dropna(axis=0, how='any', thresh=None, subset=None, inplace=False)
+    valid_trips_idx_ls = valid_trips_df.index.tolist()
+    valid_trips = [non_empty_trips[i]for i in valid_trips_idx_ls]
+
+    sim = similarity.similarity(valid_trips, radius)
+    filter_trips = sim.data
+    return filter_trips,sim,trips
+
+def valid_user(filter_trips,trips):
+    valid = False
+    if len(filter_trips) >= 10 and len(filter_trips) / len(trips) >= 0.5:
```

Comment:
if I understand your code correctly, you are determining valid users by looking at `filter_trips` which are the trips above the cutoff. Seems like you should use `valid_trips` (the trips with valid labels) instead. Not sure that will make a difference for the results, but it is also a simple fix.

Reply:
ah I see that you have not called `bin_data` so the `filter_trips` are the same as the `valid_trips`. But then why pass it in to similarity in the first place?

Reply:
`filter_trips` are just valid trips after filtering out the points, not trips above cutoff.
All trips that are used for binning and clustering are based on `filter_trips`, `trips` are original trips, so valid users come from here.

Reply:
I saw that 

> ah I see that you have not called bin_data so the filter_trips are the same as the valid_trips. But then why pass it in to similarity in the first place?

But without a comment about the "filtering out the points" to explain what you are doing, the reason for this is unclear. Please add the comment.